File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Model/Expression/ArithmeticExpression.java
package Model.Expression;

import Util.Dictionary.MyIDictionary;
import Util.Heap.IHeap;

public class ArithmeticExpression implements IExpression {
    private char operator;
    private IExpression operand1,operand2;

    public ArithmeticExpression(char operator, IExpression operand1, IExpression operand2){
        this.operator = operator;
        this.operand1 = operand1;
        this.operand2 = operand2;
    }

    @Override
    public int evaluate(MyIDictionary<String,Integer> symbolTable,IHeap heap) throws ExpressionException{
        int first = this.operand1.evaluate(symbolTable,heap);
        int second = this.operand2.evaluate(symbolTable,heap);

        switch(this.operator) {
            case '+': return first + second;
            case '-': return first - second;
            case '*': return first * second;
            case '/':
                if(second == 0)
                    throw new ExpressionException("Division by 0!");
                return first / second;
            default: throw new ExpressionException("Invalid operator!");
        }
    }

    @Override
    public String toString() {
        return "(" + operand1 + operator + operand2 + ")";
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Model/Expression/ArithmeticExpression.cs
using ToyLanguage.Util.Dictionary;

namespace ToyLanguage.Model.Expression
{
    class ArithmeticExpression : IExpression
    {
        private readonly IExpression operand1;
        private readonly IExpression operand2;
        private readonly char operatorSymbol;

        public ArithmeticExpression(char operatorSymbol,IExpression operand1,IExpression operand2)
        {
            this.operatorSymbol = operatorSymbol;
            this.operand1 = operand1;
            this.operand2 = operand2;
        }

        public int Evaluate(MyIDictionary<string, int> symbolTable)
        {
            int first = operand1.Evaluate(symbolTable);
            int second = operand2.Evaluate(symbolTable);

            switch (operatorSymbol)
            {
                case '+': return first + second;
                case '-': return first - second;
                case '*': return first * second;
                case '/':
                if (second == 0)
                    throw new ExpressionException("Division by 0!");
                return first / second;
                default:
                    throw new ExpressionException("Invalid operator!");
            }
        }

        public override string ToString()
        {
            return "ArithmeticExpression(" + operand1 + " " + operatorSymbol + " " + operand2 + ")";
        }
    }
}
